### ISSUE TITLE
Skip unused_imports check when syntax error is detected

### ISSUE THREAD

**Issue by ds-cbo** [rust-lang/rust#115892]
### Code

```Rust
mod test {
    pub trait Foo {
        fn bar(&self) {}
    }

    impl Foo for u8 { }
}


// warning: unused import: `test::Foo`
use test::Foo;

fn main() {
    // error: expected one of ..., found `oopsy`
    an oopsy;

    42_u8.bar();
}


https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=43c8bd0d4366e12e3e3c388447191605
```

### Current output

```Shell
Compiling playground v0.0.1 (/playground)
error: expected one of `!`, `.`, `::`, `;`, `?`, `{`, `}`, or an operator, found `oopsy`
  --> src/main.rs:15:8
   |
15 |     an oopsy;
   |        ^^^^^ expected one of 8 possible tokens

warning: unused import: `test::Foo`
  --> src/main.rs:11:5
   |
11 | use test::Foo;
   |     ^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: `playground` (bin "playground") generated 1 warning
error: could not compile `playground` (bin "playground") due to previous error; 1 warning emitted
```

### Desired output

```Shell
Compiling playground v0.0.1 (/playground)
error: expected one of `!`, `.`, `::`, `;`, `?`, `{`, `}`, or an operator, found `oopsy`
  --> src/main.rs:15:8
   |
15 |     an oopsy;
   |        ^^^^^ expected one of 8 possible tokens

error: could not compile `playground` (bin "playground") due to previous error
```

### Rationale and extra context

When debugging my code, I tend to start with resolving the issues at the bottom of the compiler output, since that's where my terminal naturally lands. After following the suggestion of removing the unused import, I continue with resolving the syntax error. When that's done, however, I end up re-adding the removed import because it was necessary after all.

**Comment by chenyukang** [comment]
The error message for the root cause will almost come from the top to bottom, I believe other compilers follow the same order.

We generate errors on the fly in the compiling phase, so it's also nature in this order, so I guess it is not necessary to fix it.
